pkg/controller/registry/resolver/evolver.go

```diff
@@ -153,7 +162,65 @@ func (e *NamespaceGenerationEvolver) queryForRequiredAPIs() error {
 	return nil
 }
 
+func (e *NamespaceGenerationEvolver) downgradeUpdates() error {
+	// no need to attempt downgrades
+	if len(e.gen.MissingAPIs()) == 0 {
+		return nil
+	}
+
+	// smart downgrades are only supported if fewer than 64 updates are resolved at the same time
+	// (this should be all but pathological cases)
+	if len(e.replacements) > 64 {
+		return nil
+	}
+
+	// updates[1][x] = old
+	// updates[0][x] = new
```

Comment:
Definitely non blocking, but I'm curious: This feels a little odd to reason about. Why not just have two slices? The bitwise operation is very nice, but do we save more than the cost of the cognitive load here -- especially given that we are restricting this to a relatively small constraint set?

Reply:
updated to your suggestion